pkg/controller/seed-controller-manager/cluster-stuck-controller/controller.go

```diff
@@ -0,0 +1,117 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterstuckcontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName = "kkp-cluster-stuck-controller"
+)
+
+type Reconciler struct {
+	ctrlruntimeclient.Client
+
+	workerName string
+	recorder   record.EventRecorder
+	log        *zap.SugaredLogger
+}
+
+// Add creates a new cluster-stuck controller.
+func Add(mgr manager.Manager, numWorkers int, workerName string, log *zap.SugaredLogger) error {
+	reconciler := &Reconciler{
+		Client: mgr.GetClient(),
+
+		workerName: workerName,
+		recorder:   mgr.GetEventRecorderFor(ControllerName),
+		log:        log,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to create watch: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("cluster", request.Name)
+	log.Debug("Reconciling")
+
+	// If the seed-ctrl-mgr is itself controlled by a worker, it should
+	// not influence non-workered clusters.
+	if r.workerName != "" {
+		return reconcile.Result{}, nil
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		return reconcile.Result{}, ctrlruntimeclient.IgnoreNotFound(err)
+	}
+
+	// not deleted
+	if cluster.DeletionTimestamp == nil {
+		return reconcile.Result{}, nil
+	}
+
+	// paused clusters will not be removed
+	if cluster.Spec.Pause {
+		reason := cluster.Spec.PauseReason
+		if reason == "" {
+			reason = "no reason given"
+		}
+
+		r.recorder.Eventf(cluster, corev1.EventTypeWarning, "ClusterPaused", "Cluster is paused: %s", reason)
```

Comment:
Nit: maybe we could also mention that the cluster cannot be deleted because of this (like as we do for `worker-name`).